datasource/keyvaultsecret/data.go

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+//go:generate packer-sdc struct-markdown
+//go:generate packer-sdc mapstructure-to-hcl2 -type Config,DatasourceOutput
+
+package keyvaultsecret
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/hashicorp/hcl/v2/hcldec"
+	"github.com/hashicorp/packer-plugin-azure/builder/azure/common/log"
+	"github.com/hashicorp/packer-plugin-sdk/common"
+	"github.com/hashicorp/packer-plugin-sdk/hcl2helper"
+	packersdk "github.com/hashicorp/packer-plugin-sdk/packer"
+	"github.com/hashicorp/packer-plugin-sdk/template/config"
+
+	"github.com/zclconf/go-cty/cty"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/azure-sdk-for-go/sdk/keyvault/azsecrets"
+)
+
+type Config struct {
+	common.PackerConfig `mapstructure:",squash"`
+
+	// The name of the Azure Key Vault.
+	VaultName string `mapstructure:"vault_name" required:"true"`
+	// The name of the secret to fetch from the Azure Key Vault.
+	SecretName string `mapstructure:"secret_name" required:"true"`
+	// The version of the secret to fetch. If not provided, the latest version will be used.
+	Version string `mapstructure:"version"`
+
+	// Optional fields for authentication.
+	// If not provided, the DefaultAzureCredential will be used.
+	// This includes environment variables, managed identity, etc.
+
+	TenantID     string `mapstructure:"tenant_id"`
+	ClientID     string `mapstructure:"client_id"`
+	ClientSecret string `mapstructure:"client_secret"`
+}
+
+type Datasource struct {
+	config Config
+}
+
+type DatasourceOutput struct {
+	// The raw string response of the secret version.
+	Response string `mapstructure:"response"`
+
+	// The value extracted using the 'key', if provided.
+	Value string `mapstructure:"value"`
+}
+
+func (d *Datasource) ConfigSpec() hcldec.ObjectSpec {
+	return d.config.FlatMapstructure().HCL2Spec()
+}
+
+func (d *Datasource) OutputSpec() hcldec.ObjectSpec {
+	return (&DatasourceOutput{}).FlatMapstructure().HCL2Spec()
+}
+
+func (d *Datasource) Configure(raws ...interface{}) error {
+	err := config.Decode(&d.config, nil, raws...)
+	if err != nil {
+		return err
+	}
+
+	var errs *packersdk.MultiError
+
+	if d.config.VaultName == "" {
+		errs = packersdk.MultiErrorAppend(errs, errors.New("a 'vault_name' must be specified"))
+	}
+	if d.config.SecretName == "" {
+		errs = packersdk.MultiErrorAppend(errs, errors.New("a 'secret_name' must be specified"))
+	}
+
+	if d.config.TenantID == "" {
+		d.config.TenantID = os.Getenv("AZURE_TENANT_ID")
+	}
+	if d.config.ClientID == "" {
+		d.config.ClientID = os.Getenv("AZURE_CLIENT_ID")
+	}
+	if d.config.ClientSecret == "" {
+		d.config.ClientSecret = os.Getenv("AZURE_CLIENT_SECRET")
```

Comment:
I assume that AZURE_TENANT_ID/etc are environment variables that the Azure SDK uses, since they aren't used elsewhere in the plugin this behavior may be confusing, we should document it clearly, but preferably we wouldn't use one type of authentication setup in one Packer Azure resource that we aren't using in the other builders